---
layout: post
#标题配置
title: BSseq validation report
#时间配置
date:   2019-03-27 16:36:45 +0800
#大类配置
categories: document
#小类配置
tag: 教程
---

* content
{:toc}

## WGBS分析流程测试报告

__测试人员：__ xliu@diagnoa
__测试时间：__ 2019/02/02

### 1. 模拟数据

#### 1.1 下载序列

```bash
# "Download BS-seq sequencing data for base quality simulation "
fastq-dump  --split-files --gzip ERR2676769
fastq-dump  --split-files --gzip ERR2676771
# "Reference genome FASTA file, ftp://ftp.broadinstitute.org/bundle/hg19"
reference=/mnt/hd02.d/ldu.d/software.d/VariantCall.d/sentieon/quick_start.d/reference/ucsc.hg19_chr22.fasta
```

#### 1.2 生成模拟数据

DNemulator is a package for simulating DNA sequencing errors, polymorphisms, cytosine methylation and bisulfite conversion. The [bisulfite alignment benchmark](http://cbrc3.cbrc.jp/~martin/dnemulator/bab/) provides an example of using DNemulator.

```bash
### simulation setting
export PATH=/home/xmliu/software/genome/bin/:$PATH

fasta-methyl-simv1 /mnt/hd02.d/ldu.d/software.d/VariantCall.d/sentieon/quick_start.d/reference/ucsc.hg19_chr22.fasta > meth_cancer.fa
fasta-polymorph -a poly_cancer.seg /mnt/hd03.d/xmliu.d/database.d/snp132Common_chr22.txt meth_cancer.fa meth_cancer.fa > poly_cancer.fa
fasta-paired-chunks -n 10000000 -l 126  poly_cancer.fa c5_tumor_1.fa c5_tumor_2.fa

fasta-bisulf-sim c5_tumor_1.fa > c5_tumor_1.fasta
fasta-bisulf-sim c5_tumor_2.fa > c5_tumor_2.fasta

fastq-sim c5_tumor_1.fasta <(zcat /mnt/hd03.d/xmliu.d/data.d/Bisulphite-seq/C5_colon_cancer/ERR2676771_1.fastq.gz) | gzip -c -> c5_tumor_1.fastq.gz
fastq-sim c5_tumor_2.fasta <(zcat /mnt/hd03.d/xmliu.d/data.d/Bisulphite-seq/C5_colon_cancer/ERR2676771_2.fastq.gz) | gzip -c -> c5_tumor_2.fastq.gz
```

#### 1.3 预期结果
```bash
### Generate true cpg methylation sites: /mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation/poly_normal.cpg_sites
python /mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation/cpgsites.py 
```
### 2. 测试流程

#### 2.1 Bismark流程

```bash
#### basic settings
db="/mnt/hd04.d/xmliu.d/project.d/methylation/proj_db"
workdir="$PWD/bs_test_bismark"
raw_data="/mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation"

mkdir -p $workdir && cd $workdir && mkdir -p s01_fastqc s02_alignment/s020_bismark s02_alignment/s021_dedup s03

#### step1 fastQC && trim-glora
trim_galore --fastqc --paired --clip_R1 4 --clip_R2 4 -o $workdir/s01_fastqc/ $raw_data/c5_tumor_1.fastq.gz $raw_data/c5_tumor_2.fastq.gz

#### step2.1 biamark mapping and post algnment process
bismark --bowtie2 -N 1 -L 20 --score_min L,0,-1.6 $db -1 $workdir/s01_fastqc/c5_tumor_1_val_1.fq.gz -2 $workdir/s01_fastqc/c5_tumor_2_val_2.fq.gz  -o $workdir/s02_alignment/s020_bismark  &&\
            samtools sort -m 1G $workdir/s02_alignment/s020_bismark/c5_tumor_1_val_1_bismark_bt2_pe.bam -o $workdir/s02_alignment/s020_bismark/c5_tumor_1_val_1_bismark_bt2_pe_sorted.bam &&\
            deduplicate_bismark -p --bam $workdir/s02_alignment/s020_bismark/c5_tumor_1_val_1_bismark_bt2_pe_sorted.bam -o $workdir/s02_alignment/s021_dedup

#### step3.1 methylation extractor
            bismark_methylation_extractor --output $workdir/s03/tumor --gzip --bedGraph $workdir/s02_alignment/s021_dedup/c5_tumor_1_val_1_bismark_bt2_pe_sorted.deduplicated.bam

            ```
#### 2.2 Bwa-meth流程
            ```
            ref="/mnt/hd04.d/xmliu.d/project.d/methylation/proj_db/ucsc.hg19_chr22.fasta"
            workdir="$PWD/bs_test_bwameth"
            raw_data="/mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation"

            mkdir -p $workdir && cd $workdir && mkdir -p s01_fastqc s02_alignment/s020_bwameth  s02_alignment/s021_dedup s03

#### step1 fastQC && trim-glora
            trim_galore --fastqc --paired --clip_R1 4 --clip_R2 4 -o $workdir/s01_fastqc/ $raw_data/c5_tumor_1.fastq.gz $raw_data/c5_tumor_2.fastq.gz

#### step2 bwa-meth mapping and post algnment process
            python /home/xmliu/software/basic_bin/miniconda2/bin/bwameth.py  -t 4 --reference $ref --read-group '@RG\tID:tumor\tSM:tumor' $workdir/s01_fastqc/c5_tumor_1_val_1.fq.gz $workdir/s01_fastqc/c5_tumor_2_val_2.fq.gz | samtools view -b -> $workdir/s02_alignment/s020_bwameth/c5_tumor.bam
            samtools sort $workdir/s02_alignment/s020_bwameth/c5_tumor.bam -o $workdir/s02_alignment/s020_bwameth/c5_tumor_sorted.bam && samtools index $workdir/s02_alignment/s020_bwameth/c5_tumor_sorted.bam

#### step3 methylation extractor
            MethylDackel extract $ref $workdir/s02_alignment/s020_bwameth/c5_tumor_sorted.bam -o $workdir/s03/c5_tumor
            ```

#### 2.3 BSmap流程
            ```
            ref="/mnt/hd04.d/xmliu.d/project.d/methylation/proj_db/ucsc.hg19_chr22.fasta"
            workdir="$PWD/bs_test_bsmap"
            raw_data="/mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation"
            export PATH=/home/xmliu/software/alignment_algorithm/bsmap-2.90/samtools/:$PATH

            mkdir -p $workdir && cd $workdir && mkdir -p s01_fastqc s02_alignment/s020_bsmap  s02_alignment/s021_dedup s03

#### step1 fastQC && trim-glora
            trim_galore --fastqc --paired --clip_R1 4 --clip_R2 4 -o $workdir/s01_fastqc/ $raw_data/c5_tumor_1.fastq.gz $raw_data/c5_tumor_2.fastq.gz

#### step2 bsmap  mapping and post algnment process
            bsmap -a $workdir/s01_fastqc/c5_tumor_1_val_1.fq.gz -b $workdir/s01_fastqc/c5_tumor_2_val_2.fq.gz -d $ref -v 1 -w 2 -r 1 -n 1 -m 0 -x 1000 -o $workdir/s02_alignment/s020_bsmap/c5_tumor_bsmap.bam
            samtools sort $workdir/s02_alignment/s020_bsmap/c5_tumor_bsmap.bam  $workdir/s02_alignment/s020_bsmap/c5_tumor_bsmap_sorted && samtools index $workdir/s02_alignment/s020_bsmap/c5_tumor_bsmap_sorted.bam

#### step3 methylation extractor
            python /home/xmliu/software/alignment_algorithm/bsmap-2.90/methratio.py -d $ref  $workdir/s02_alignment/s020_bsmap/c5_tumor_bsmap_sorted.bam  > $workdir/s03/c5_tumor_bsmap
            ```

#### 2.4 实际结果
##### Methylation cpg_sites:
            ```
            /mnt/hd04.d/xmliu.d/project.d/methylation/analysis/bs_test_bismark/s03/tumor/c5_tumor_1_val_1_bismark_bt2_pe_sorted.deduplicated.bedGraph
            /mnt/hd04.d/xmliu.d/project.d/methylation/analysis/bs_test_bwameth/s03/c5_tumor_CpG.bedGraph
            /mnt/hd04.d/xmliu.d/project.d/methylation/analysis/bs_test_bsmap/s03/c5_tumor_bsmap_cpg_sites.txt
            ```
### 3. 统计分析

#### 3.1 bismark 
            ``` bismark
            python3 /home/xmliu/bin/evaluate_bed.py   /mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation/poly_cancer.cpg_sites.CpG.txt   /mnt/hd04.d/xmliu.d/project.d/methylation/analysis/bs_test_bismark/s03/tumor/c5_tumor_1_val_1_bismark_bt2_pe_sorted.deduplicated.bedGraph
            ```
            class|True Postive|False Postive|False Negative
            ----|----|----|----
            total no.|990005 |114411| 50280

            Header|sensitivity |  precision |  F1-score
            ----|----|----|----
            ratio|0.9516 | 0.8964 | 0.9232

#### 3.2 bwa-meth
            ``` bwa-meth
            python3  /home/xmliu/bin/valuate_bed.py   /mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation/poly_cancer.cpg_sites.CpG.txt   /mnt/hd04.d/xmliu.d/project.d/methylation/analysis/bs_test_bwameth/s03/c5_tumor_CpG.bedGraph
            ```
            class|True Postive|False Postive|False Negative
            ----|----|----|----
            total no.|1012532| 112811| 27753

            Header|sensitivity |  precision |  F1-score
            ----|----|----|----
            ratio|0.9733|0.8998|  0.9351

#### 3.3 BSmap
            ``` BSmap
            python3  /home/xmliu/bin/valuate.py   /mnt/hd04.d/xmliu.d/project.d/methylation/raw_data/simulation/poly_cancer.cpg_sites.CpG.txt   /mnt/hd04.d/xmliu.d/project.d/methylation/analysis/bs_test_bsmap/s03/c5_tumor_bsmap_cpg_sites.txt
            ```
            class|True Postive|False Postive|False Negative
            ----|----|----|----
            total no.|1036443|115466|3842

            Header|sensitivity |  precision |  F1-score
            ----|----|----|----
            ratio|0.9963| 0.8998|0.9456
### 4. 总结
            在本次测试中，三个甲基化比对软件均可以保证较高的灵敏度和准确性。
            其中bismark和bwa-meth都是基于3-letter的原理，而BSmap则是基于wildtype的原理。三款软件中bismark运行时间最长，但是占用内存较少；bwa-meth和BSmap运行时间更短，但是内存资源占用较多。
            参数设置方面，bwa-meth最为友好，一般使用默认参数就可以达到比较理想的比对。在本次使用DnNemulator生成的模拟数据上，bismark默认参数的比对结果很不理想（pe mapping rate < 3%），仔细调整参数之后才到达预期效果，使用时需注意测序数据类型和质量以调整参数。

